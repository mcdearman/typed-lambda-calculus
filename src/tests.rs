use chumsky::{input::Stream, prelude::Input, span::SimpleSpan, Parser};
use logos::Logos;

use crate::{parser, Token};

#[test]
fn test_parse_let() {
    let src = "let x = 5 in x";
    let lex = Token::lexer(&src)
        .spanned()
        .map(|(tok, span)| (tok, SimpleSpan::from(span)));
    let tok_stream = Stream::from_iter(lex).spanned(SimpleSpan::from(src.len()..src.len()));
    let ast = parser()
        .parse(tok_stream)
        .into_result()
        .expect("failed to parse");
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn test_parse_multiline_let() {
    let src = "let x =\n5 in x";
    let lex = Token::lexer(&src)
        .spanned()
        .map(|(tok, span)| (tok, SimpleSpan::from(span)));
    let tok_stream = Stream::from_iter(lex).spanned(SimpleSpan::from(src.len()..src.len()));
    let ast = parser()
        .parse(tok_stream)
        .into_result()
        .expect("failed to parse");
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn test_parse_apply() {
    let src = "let f = \\x -> x in f 5";
    let lex = Token::lexer(&src)
        .spanned()
        .map(|(tok, span)| (tok, SimpleSpan::from(span)));
    let tok_stream = Stream::from_iter(lex).spanned(SimpleSpan::from(src.len()..src.len()));
    let ast = parser()
        .parse(tok_stream)
        .into_result()
        .expect("failed to parse");
    insta::assert_debug_snapshot!(ast);
}

#[test]
fn test_parse_multiline_apply() {
    let src = "let f =\n\\x -> x in f 5";
    let lex = Token::lexer(&src)
        .spanned()
        .map(|(tok, span)| (tok, SimpleSpan::from(span)));
    let tok_stream = Stream::from_iter(lex).spanned(SimpleSpan::from(src.len()..src.len()));
    let ast = parser()
        .parse(tok_stream)
        .into_result()
        .expect("failed to parse");
    insta::assert_debug_snapshot!(ast);
}
