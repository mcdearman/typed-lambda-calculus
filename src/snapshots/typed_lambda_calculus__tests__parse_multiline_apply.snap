---
source: src/tests.rs
expression: ast
---
Let(
    Ident(f),
    Lambda(
        Ident(x),
        Var(
            Ident(x),
        ),
    ),
    Apply(
        Var(
            Ident(f),
        ),
        Int(
            5,
        ),
    ),
)
